br/pkg/lightning/restore/check_info.go

```diff
@@ -650,86 +652,257 @@ func (rc *Controller) SchemaIsValid(ctx context.Context, tableInfo *mydump.MDTab
+			continue
+		}
+		_, row, err := rc.readFirstRow(ctx, f.FileMeta)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(row) > 0 {
+			rows = append(rows, row)
+		}
+		// only check at most two of all the files
+		if len(rows) >= 2 {
+			break
+		}
+	}
+	if len(rows) == 0 {
+		return nil
+	} else if len(rows) >= 2 {
+		// if the first row in two source files are not the same, they should not be the header line
+		// NOTE: though lightning's logic allows different source files contains different columns or the
+		// order is difference, here we only check if they are exactly the same because this is the common case.
+		if len(rows[0]) != len(rows[1]) {
+			return nil
+		}
+
+		for i := 0; i < len(rows[0]); i++ {
+			if rows[0][i].GetString() != rows[1][i].GetString() {
+				return nil
+			}
+		}
+	}
+
+	// check if some fields are unique and not ignored
+	// if at least one field appears in a unique key, we can sure there is something wrong,
+	// they should be either the header line or the data is duplicated.
+	tableInfo := rc.dbInfos[tableMeta.DB].Tables[tableMeta.Name]
+	tableFields := make(map[string]struct{})
+	uniqueIdxFields := make(map[string]struct{})
+	ignoreColumns, err := rc.cfg.Mydumper.IgnoreColumns.GetIgnoreColumns(tableMeta.DB, tableMeta.Name, rc.cfg.Mydumper.CaseSensitive)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	ignoreColsSet := make(map[string]struct{})
+	for _, col := range ignoreColumns.Columns {
+		ignoreColsSet[col] = struct{}{}
+	}
+	for _, idx := range tableInfo.Core.Indices {
+		if !idx.Unique && !idx.Primary {
+			continue
+		}
+		for _, col := range idx.Columns {
+			if _, ok := ignoreColsSet[col.Name.L]; !ok {
+				uniqueIdxFields[col.Name.L] = struct{}{}
+			}
+		}
+	}
+	for _, f := range tableInfo.Core.Columns {
+		tableFields[f.Name.L] = struct{}{}
+	}
+	if common.TableHasAutoRowID(tableInfo.Core) {
+		tableFields[model.ExtraHandleName.L] = struct{}{}
+	}
+	hasUniqueField := false
+	for _, d := range rows[0] {
+		val := strings.ToLower(d.GetString())
+		if _, ok := tableFields[val]; !ok {
+			return nil
+		}
+		if _, ok := uniqueIdxFields[val]; ok {
+			hasUniqueField = true
+			break
+		}
+	}
+
+	msg := fmt.Sprintf("source csv files contains header row but `mydumper.csv.header` is false, checked table is `%s`.`%s`",
+		tableMeta.DB, tableMeta.Name)
+	level := Warn
+	if hasUniqueField && len(rows) > 1 {
+		level = Critical
+	}
+	// if there are only 1 csv file or there is not unique key, try to check if all columns are compatible with string value
+	if !checkFieldCompatibility(tableInfo.Core, rows[0]) {
+		level = Critical
+	}
+	rc.checkTemplate.Collect(level, false, msg)
+
+	return nil
+}
+
+func checkFieldCompatibility(tbl *model.TableInfo, values []types.Datum) bool {
+	se := kv.NewSession(&kv.SessionOptions{
+		SQLMode: mysql.ModeStrictTransTables,
+	})
+	for i, col := range tbl.Columns {
+		if i >= len(values) {
+			break
+		}
+		_, err := table.CastValue(se, values[i], col, true, false)
```

Comment:
If there are some ignore columns, does values[i] still match tbl.Columns[i]?

Reply:
When `csv.header = false`, there is a pre-assumption that columns in source files are the same as table schema, so ignore-columns doesn't change this prerequest. But It's still better to skip check the ignored-columns.